Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs

```diff
@@ -0,0 +1,131 @@
+
+        // Raise an event to get any modifiers
+        var modifierEvent = new RefreshNameModifiersEvent(entity.Comp?.BaseName ?? meta.EntityName);
+        RaiseLocalEvent(entity, ref modifierEvent);
+
+        // No modifiers
+        if (modifierEvent.ModifierCount == 0)
+        {
+            // If the entity doesn't have the component, we're done
+            if (entity.Comp == null)
+                return;
+
+            // Restore the base name
+            _metaData.SetEntityName(entity, entity.Comp.BaseName, meta, raiseEvents: false);
+            // The component isn't doing anything anymore, so remove it
+            RemComp<NameModifierComponent>(entity);
+            return;
+        }
+
+        // Get the final name with modifiers applied
+        var modifiedName = modifierEvent.GetModifiedName();
+
+        // Add the component if needed, and store the base name
+        if (!EnsureComp<NameModifierComponent>(entity, out var comp))
+            SetBaseName((entity, comp), meta.EntityName);
+
+        // Set the entity's name with modifiers
+        _metaData.SetEntityName(entity, modifiedName, meta, raiseEvents: false);
+    }
+}
+
+/// <summary>
+/// Raised on an entity when <see cref="NameModifierSystem.RefreshNameModifiers"/> is called.
+/// Subscribe to this event and use its methods to add modifiers to the entity's name.
+/// </summary>
+[ByRefEvent]
+public sealed class RefreshNameModifiersEvent : IInventoryRelayEvent
+{
+    /// <summary>
+    /// The entity's name without any modifiers applied.
+    /// If you want to base a modifier on the entity's name, use
+    /// this so you don't include other modifiers.
+    /// </summary>
+    public readonly string BaseName;
+
+    private readonly List<(LocId LocId, int Priority, ValueList<(string, object)>? ExtraArgs)> _modifiers = [];
+
+    /// <inheritdoc/>
+    public SlotFlags TargetSlots => ~SlotFlags.POCKET;
+
+    public int ModifierCount => _modifiers.Count;
+
+    public RefreshNameModifiersEvent(string baseName)
+    {
+        BaseName = baseName;
+    }
+
+    /// <summary>
+    /// Adds a modifier to the entity's name.
+    /// The original name will be passed to Fluent as <c>$baseName</c> along with any <paramref name="extraArgs"/>.
+    /// Modifiers with a higher <paramref name="priority"/> will be applied later.
+    /// </summary>
+    public void AddModifier(LocId locId, int priority = 0, ValueList<(string, object)>? extraArgs = null)
+    {
+        _modifiers.Add((locId, priority, extraArgs));
+    }
+
+    /// <summary>
+    /// Returns the final name with all modifiers applied.
+    /// </summary>
+    public string GetModifiedName()
+    {
+        var name = BaseName;
+
+        foreach (var modifier in _modifiers.OrderBy(n => n.Priority))
+        {
+            var args = modifier.ExtraArgs ?? [];
+            args.Add(("baseName", name));
+            name = Loc.GetString(modifier.LocId, args.ToArray());
+        }
```

Comment:
this is kinda odd, if this is supposed to always override the lesser priority modifiers with the higher priority ones, couldnt you just order by priority (extra points without linq) and just use the last element of the sorted list instead of unnecessarily looping?

Reply:
This is a little tricky - it needs to loop so that modifiers that work as a prefix/postfix can be applied together.

So if you have a metamorphic glass with beer in it and you put glue, lube, and a label on it, the different systems will add `glued {$baseName}` (priority 0),  `lubed {$baseName}` (priority 0), `beer glass` (priority -1), and `{$baseName} (label)`

This code loops through like: 
base name: "metamorphic glass"
`beer glass` -> "beer glass"
`glued {$baseName}` -> "glued beer glass"
`lubed {$baseName}` -> "lubed glued beer glass"
`{$baseName} (label)` -> "lubed glued beer glass (label)"

So the loop is necessary. I definitely can add some comments to make the logic clearer for future maintainers.

Reply:
Ah no i kinda got that, i think what i misunderstood is what _modifiers was actually getting populated with, kinda conflated it with the extra args, my bad